File: app/src/main/java/com/commit451/gitlab/adapter/BranchAdapter.java
package com.commit451.gitlab.adapter;

import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import com.commit451.gitlab.R;
import com.commit451.gitlab.model.Ref;
import com.commit451.gitlab.model.api.Branch;
import com.commit451.gitlab.viewHolder.BranchViewHolder;

import java.util.ArrayList;
import java.util.Collection;

/**
 * Adapts the feeds
 */
public class BranchAdapter extends RecyclerView.Adapter<BranchViewHolder> {


    private Listener listener;

    private ArrayList<Branch> values;
    @Nullable
    private Ref ref;

    public BranchAdapter(@Nullable Ref currentRef, Listener listener) {
        this.listener = listener;
        values = new ArrayList<>();
        ref = currentRef;
    }

    private final View.OnClickListener mOnItemClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            int position = (int) v.getTag(R.id.list_position);
            listener.onBranchClicked(getEntry(position));
        }
    };

    public void setEntries(Collection<Branch> entries) {
        values.clear();
        if (entries != null) {
            values.addAll(entries);
        }
        notifyDataSetChanged();
    }

    @Override
    public BranchViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        BranchViewHolder holder = BranchViewHolder.inflate(parent);
        holder.itemView.setOnClickListener(mOnItemClickListener);
        return holder;
    }

    @Override
    public void onBindViewHolder(final BranchViewHolder holder, int position) {
        holder.itemView.setTag(R.id.list_position, position);
        Branch branch = getEntry(position);
        boolean selected = false;
        if (ref != null) {
            if (ref.getType() == Ref.TYPE_BRANCH
                    && ref.getRef().equals(branch.getName())) {
                selected = true;
            }
        }
        holder.bind(branch, selected);
    }

    @Override
    public int getItemCount() {
        return values.size();
    }

    private Branch getEntry(int position) {
        return values.get(position);
    }

    public interface Listener {
        void onBranchClicked(Branch entry);
    }
}

File: app/src/main/java/com/commit451/gitlab/adapter/BranchAdapter.kt
package com.commit451.gitlab.adapter

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import com.commit451.gitlab.R
import com.commit451.gitlab.model.Ref
import com.commit451.gitlab.model.api.Branch
import com.commit451.gitlab.viewHolder.BranchViewHolder
import java.util.*

/**
 * Adapts the feeds
 */
class BranchAdapter(private val ref: Ref?, private val listener: BranchAdapter.Listener) : RecyclerView.Adapter<BranchViewHolder>() {

    private val values: ArrayList<Branch> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BranchViewHolder {
        val holder = BranchViewHolder.inflate(parent)
        holder.itemView.setOnClickListener { v ->
            val position = v.getTag(R.id.list_position) as Int
            listener.onBranchClicked(getEntry(position))
        }
        return holder
    }

    override fun onBindViewHolder(holder: BranchViewHolder, position: Int) {
        holder.itemView.setTag(R.id.list_position, position)
        val branch = getEntry(position)
        var selected = false
        if (ref != null) {
            if (ref.type == Ref.TYPE_BRANCH && ref.ref == branch.name) {
                selected = true
            }
        }
        holder.bind(branch, selected)
    }

    override fun getItemCount(): Int {
        return values.size
    }

    fun setEntries(entries: Collection<Branch>?) {
        values.clear()
        if (entries != null) {
            values.addAll(entries)
        }
        notifyDataSetChanged()
    }

    fun addEntries(entries: Collection<Branch>) {
        if (!entries.isEmpty()) {
            val start = values.size
            this.values.addAll(entries)
            notifyItemRangeChanged(start, this.values.size)
        }
    }

    private fun getEntry(position: Int): Branch {
        return values[position]
    }

    interface Listener {
        fun onBranchClicked(entry: Branch)
    }
}
